### ISSUE TITLE
Multiple artifactIds for the same package

### ISSUE THREAD

**Issue by manolo** [webjars/webjars#1440]
Go to http://www.webjars.org/bower and look for paper-input. You get 3 artifactIds for the same package.

Two of them are because of uppercase/lowercase issue #1438, but the other case is because there is no check against whether the package is published in bower. It's not the same ask for publishing `iron-input` than `polymerelements/iron-input`

**Comment by jamesward** [comment]
Unfortunately we need artifacts based on both the GitHub repo and based on the Bower artifact.  This is because dependency definitions can either specify a GitHub repo or a Bower artifact.

**Comment by manolo** [comment]
I understand both ways to specify the package. Nevertheless bower always considers both packages  as the same. When you try to install both your local repo `bower install iron-input polymerelements/iron-input`, only one would installed, cached, and versioned as the same package.
Since `bower info` gives you the repo of the package, based on it, it would be possible to set the correct package name, and be aligned with the bower behavior.

```
$ bower info iron-input
resolved https://github.com/PolymerElements/iron-input.git#1.0.10
```

**Comment by jamesward** [comment]
We could convert all of the short names to the URL form but I think that'd sacrifice usability.  There might be a way to convert URLs to short names that'd work.  But why do people use the git or GitHub form when there is a short name?  Is it just to be able to specify a commit sha instead of a tag?

**Comment by manolo** [comment]
The problem is that transitive dependencies defined in bower.json files are defined with the github.com form. Then this dependencies are copied to the pom.xml.

```
$ bower info iron-icon
...
  dependencies: {
    'iron-flex-layout': 'polymerelements/iron-flex-layout#^1.0.0',
    'iron-meta': 'polymerelements/iron-meta#^1.0.0',
    polymer: 'Polymer/polymer#^1.1.0'
  },
...
```

```
$ cat .../iron-icon-1.0.9.pom
     ...
        <dependency>
            <groupId>org.webjars.bower</groupId>
            <artifactId>github-com-polymer-polymer</artifactId>
            <version>[1.1.0,2)</version>
        </dependency>
    ...
```

As you can see, although I can request to deploy the short form `iron-icon` and `polymer` instead of `polymerelemens/iron-icon` and `polymer/polymer`, I always need to be deployed the long form in order to make maven dependencies work. Obviously can set a lot of exclusions and dependencies in my `pom.xml`, but that is a tough task.

Solutions should be
1.- always use the long form
2.- when the long form is requested, try to figure out whether a short name is registered in bower, and deploy the short name

**Comment by jamesward** [comment]
The Polymer project use all GitHub references in their dependencies, right?  So I'm not seeing the issue with using only the GitHub URLs for those dependencies when deploying them to WebJars.  Maybe I'm missing something.

**Comment by manolo** [comment]
> The Polymer project use all GitHub references in their dependencies, right? 

Not always, some times it looks like `"webcomponentsjs": "^0.7.0"` other times `"webcomponentsjs": "webcomponents/webcomponentsjs#^0.7.0"`

> using only the GitHub URLs for those dependencies when deploying them to WebJars

That is an option, but then we should avoid the user to deploy without the github reference. Otherwise we end messing up pom.xml packages deployed with the short name but depending on the large form name.

IMO, webjars should follow the same criteria than bower. Bower considers always the `name` part of the dependencies object to track the package and its dependencies.

So to me, it does not make sense that a dependency defined in bower.json as `"webcomponentsjs": "webcomponents/webcomponentsjs#^0.7.0"` does not go to the pom.xml as `webcomponentsjs` 

In fact, and due to the rigid maven system, I will not allow deploying to maven central repos packages that are not correctly registered in bower. It would prevent broken packages like we have now that are irreversible.






I have spent a long time trying to make work something simple like:

      <dependency>
        <groupId>org.webjars.bower</groupId>
        <artifactId>paper-elements</artifactId>
        <version>1.0.9</version>
      </dependency>

**Comment by jamesward** [comment]
But there *is* actually a difference between `"webcomponentsjs": "^0.7.0"` and `"webcomponentsjs": "webcomponents/webcomponentsjs#^0.7.0"`.  The first is a reference to the artifact registered in the Bower repo as `webcomponentsjs` and the second is a reference to the GitHub repo `webcomponents/webcomponentsjs`.  The two might be related but might not.

We might be able to try and convert `webcomponents/webcomponentsjs` to the registered artifact named `webcomponentsjs` but there will still be instances where there isn't a registered artifact associated with a provided GitHub repo.

The fact that Bower allows this messiness means that WebJars also needs to allow it.

Can you provide and example artifact that isn't working as expected?

**Comment by manolo** [comment]
Try this one

```
      <dependency>
        <groupId>org.webjars.bower</groupId>
        <artifactId>iron-elements</artifactId>
        <version>[1.0.0,2)</version>
      </dependency>
```

Or these two together
```
      <dependency>
        <groupId>org.webjars.bower</groupId>
        <artifactId>github-com-vaadin-vaadin-combo-box</artifactId>
        <version>[1.0.0,2)</version>
      </dependency>
      <dependency>
        <groupId>org.webjars.bower</groupId>
        <artifactId>github-com-vaadin-vaadin-date-picker</artifactId>
        <version>[1.0.0,2)</version>
      </dependency>
```

**Comment by manolo** [comment]
Thinking a little bit more about this, I think that this should be a good solution:
- When deploying without the github part `iron-elements`, all dependencies in pom.xml should use the left part of the dependency definition in bower `<artifactId>iron-list</artifactId>`. That means that the user has to deploy all official bower packages for that artefact.
- Otherwise `polymerelements/iron-elements` the dependencies in pom.xml should look like currently `<artifactId>com-github-polymerelements-iron-list</artifactId>`

**Comment by jamesward** [comment]
We can't always use the left side of the dependency definition because it may not actually correspond to the provided GitHub repo.  That is why today the default is to use the GitHub repo for the name.  When creating the dependency definition we could check if the two do correspond and if so use the name on the left.

I tried the `iron-elements` Bower WebJar and there are some unresolved dependencies:
```
[warn] 	:: org.webjars.bower#github-com-PolymerElements-iron-label;[1.0.0,2): not found
[warn] 	:: org.webjars.bower#github-com-PolymerElements-iron-swipeable-container;[1.0.0,2): not found
[warn] 	:: org.webjars.bower#github-com-PolymerElements-iron-scroll-threshold;[1.0.0,2): not found
[warn] 	:: org.webjars.bower#github-com-PolymerElements-iron-location;[0.8.0,0.9): not found
[warn] 	:: org.webjars.bower#github-com-PolymerElements-iron-demo-helpers;[1.0.0,2): not found
```

Due to the change that automatically lowercases the artifact names on deployment we currently can not deploy those missing artifacts.  So I'll need to come up with some way to override that behavior so that those missing WebJars can be deployed.

**Comment by manolo** [comment]
> We can't always use the left side of the dependency definition because it may not actually correspond to the provided GitHub repo

It could not correspond, but it is responsibility of the developer have it aligned because in fact he can change the name.
Bower behaviour is that it always uses left part for  the package name, for instance you could do `my-iron-input: polymerelements/iron-input` and bower will install and cache `my-iron-input`.
So considering the left part in dependencies for deploys without path is safe.

**Comment by jamesward** [comment]
I've seen a number of instances where the GitHub repo referenced does not correspond to the Bower package.  This is usually done when a fork contains a bug fix that has not been merged into the main project.  I've also seen a number of cases where this is not Bower package for the provided GitHub repo.  This used to be the case with a lot of the Polymer dependencies.  So we can't just trust the information blindly.

**Comment by jamesward** [comment]
I was able to get `"org.webjars.bower" % "iron-elements" % "1.0.10"` to work with all of the transitive deps.  So I'm closing this issue.  There is a new issue for the artifact naming thing: #1452

**Comment by manolo** [comment]
Well, it works to me but the second time I run maven, at least in my system (mac os).
Can you try this ?

- Create a project with this `pom.xml`
```
<project>
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.example</groupId>
  <artifactId>demo</artifactId>
  <packaging>jar</packaging>
  <version>1.0-SNAPSHOT</version>

  <dependencies>
    <dependency>
      <groupId>org.webjars.bower</groupId>
      <artifactId>iron-elements</artifactId>
      <version>1.0.10</version>
    </dependency>
  </dependencies>
</project>
```

Then
```
rm -rf ~/.m2/repository/org/webjars
mvn clean dependency:tree # fails
mvn clean dependency:tree #works
```

You can see that there are mixed downloads camel and down-cased, and also in the dependency tree. That causes random local issues when the filesystem is case unsensitive. Problems become worst when you mix more elements in your project.

I think everything will get normalised when we deploy new versions of all elements, because all transitive dependencies will be down-cased. Or when you implement #1452 (btw thanks for heading this) and everything could be redeployed with correct dependency tree based on bower registered names (still could be some old elements with camel-cased dependencies).
Another option would be to clean manually maven central and bintray repos which I guess is not possible.

**Comment by manolo** [comment]
> So we can't just trust the information blindly.

My point is, that if the user just provide the package name (without github info) he trust in bower registered info. To me, it makes sense that all dependencies of that artefact also trust in bower info (because deployer does), and they should be short names.

Otherwise if the user gives the full-path, it's safer to consider always full-path for its dependencies.

**Comment by jamesward** [comment]
I got: `No versions available for org.webjars.bower:github-com-polymerelements-prism-element:jar:[1.0.0,2) within specified range`

So I've deployed the lower case artifact.

I don't think there is any way we can fix the issue of commingled case for users that are on case insensitive file systems.

**Comment by bartolom** [comment]
I also want to use webjars to manage Polymer dependecies in our projects.
The setup is promisiong but I also ran into a few problems.

I also recognized the different all lower case and camel case naming and asked the Polymer team if they could be consistent https://github.com/PolymerLabs/tedium/issues/47 
They agreed and future versions of Polymer will be more consistent once the patch is supplied.

I see the usuability of just pulling the short name "paper-input" but maybe I'm a to long time Java developer. So for our use I'm pulling in the "qualified" dependencies.

(Gradle)
```
compile 'org.webjars.bower:github-com-PolymerElements-paper-input:1.1.17'
```

I'm excluding all the ambiguous in Gradle (which are pulled from 

```
exclude module: 'github-com-polymerelements-paper-input'
```

I was hoping that once the Polymer team is consistent on their side these ugly excludes. But now I start to understand that it is acutally the webjar-pom-generator that mixes them as well.

For instance it was actually me who deployed "github-com-PolymerElements-paper-input" 1.1.17. I typed "PolymerElements/paper-input" into the self-service webjars-GUI. I figured this out to work even though the GUI suggested me "github-com-PolymerElements-paper-input", and I was happy with it.

When one looks at http://www.webjars.org/listfiles/org.webjars.bower/github-com-PolymerElements-paper-input/1.1.17 and the bower.json there shows that the Polymer team is using the full< correct `"iron-input": "PolymerElements/iron-input#^1.0.0"` (with Capital letters) but the pom.xml generated by webjar.org has lower-cased the dependencies but leaves the artifact with Capital letters

```
<artifactId>github-com-PolymerElements-paper-input</artifactId>
...
<dependency>
            <groupId>org.webjars.bower</groupId>
            <artifactId>github-com-polymerelements-iron-input</artifactId>
            <version>[1.0.0,2)</version>
        </dependency>
```

One partial solution for me would be to switch to the all-lower-case version as well. But I think this it more correct to leave it like the original JavaScript programmers have created it. 

Therefore "github-com-PolymerElements-iron-input" would be preferred

**Comment by bartolom** [comment]
The next part of my Polymer-Webjar challenge is: Now that I have all these fully qualified but at least consistent github-com-PolymerElements-paper-input dependencies I actually need to get rid of them in the URLs. Because the Polymer projects relies heavyly on relative depenencies in the Web-Component-HTML-Import statements. They use things like this in their code (from paper-input):

```
...
<link rel="import" href="../polymer/polymer.html">
<link rel="import" href="../iron-form-element-behavior/iron-form-element-behavior.html">
<link rel="import" href="../iron-input/iron-input.html">
<link rel="import" href="paper-input-behavior.html">
...
```

They simply assume that the people pulled their dependencies into a "bower_dependencies" directory. And I think this will become more used once web components gain more traction.

This was actually possible to solve with Webjars. I wrote a piece of software that gets the list form `org.webjars.WebJarAssetLocator.getFullPathIndex(Pattern, ClassLoader...)` and there locates all the pom.xml files that are created by Webjars.org. I use then the JDK STAX XML-Parser to effeciently parse and stop parsing right afer I have all the information that I need, which are in the `<artifactId>github-com-PolymerElements-paper-input</artifactId>` and `<scm><url>https://github.com/PolymerElements/paper-input</url></scm>`

At this point I can constructs a Java Map that can "translate back" `github-com-PolymerElements-paper-input` to  `paper-input`.

The last piece of this was that I need to subclass WebJarAssetLocator and override getFullPath(String, String) and remap. (getFullPath is the public method called by Spring Boot 1.4.0 which we will use)

So maybe my parsing and remapping could become an optional part of `WebJarAssetLocator` and I could pass it in the constructor. My current code looks like this.

```
public class WebJarAssetLocatorForPolymer extends WebJarAssetLocator {

	private final Function<String, Optional<String>> tweak;

	public WebJarAssetLocatorForPolymer(
			SortedMap<String, String> fullPathIndex, Function<String, Optional<String>> tweak) {
		super(fullPathIndex);
		this.tweak = tweak;
	}

	@Override
	public String getFullPath(final String webjar, final String partialPath) {
		String safe = tweak.apply(webjar).orElse(webjar);
		return super.getFullPath(safe, partialPath);
	}


}
```

Just as note, to describe what else I have also done to make Polymer work for us: 

1.) We use Spring Boot 1.4.0 which comes with Spring-Framework 4.3.2 which itself has a hook to supply its own instance of WebJarsLocator https://jira.spring.io/browse/SPR-14092

2.) I had to blacklist quite some stuff from the fullPathIndex, because the partialPath must be free of dublicates within one webjar-pom (paper-input.html vs test/paper-input.html). But this is no problem with the current WebJarAssetLocator as it has a constructor where a custom fullPathIndex can be supplied. So I have utility code like this:

```
/**
	 * Will remove various unnecessary and often ambiguous entries from the
	 * standard WebJar fullPathIndex, so they work well with Polymer.
	 * <p>
	 * Eg.: remove "/test/", "/demo/", "/classes/"
	 * <p>
	 * Also Markdown, index.html entries are removed as they are by contention
	 * documentation.
	 *
	 * @param map
	 *            will be mutated
	 */

	static void preprocessWebjarsIndex(SortedMap<String, String> map) {
		doBlacklist(map, ".html/test/", "test");
		doBlacklist(map, ".html/classes/", "classes");
		doBlacklist(map, ".html/demo/", "demo");
		doBlacklist(map, "/src/demo/", "demo sources");
		doBlacklist(map, ".md/", "md markdown files");
		doBlacklist(map, ".travis.yml/", "travis");
		doBlacklist(map, "index.html/", "index.html documentation");
		doBlacklist(map, ".scss/scss", "scss CSS files");
		doBlacklist(map, ".less/less/", "less CSS files");
		doBlacklist(map, "pom.properties/", "pom.properties");
		doBlacklist(map, "bower.json/", "bower.json");
		doBlacklist(map, "package.json/", " package.json");
...
}
	private static boolean containsGithubComPolymer(String k) {
		return k.contains("/github-com-Polymer")
				|| k.contains("/font-awesome/")
				|| k.contains("github-com-webcomponents-webcomponentsjs");
	}

	private static void doBlacklist(SortedMap<String, String> map, String contains, String info) {
		List<String> blackListTest = map.keySet().stream()
				.filter(k -> containsGithubComPolymer(k))
				.filter(k -> k.contains(contains))
				.collect(toList());

		LOG.info("{} '{}' Polymer resources removed.", blackListTest.size(), info);
		blackListTest.forEach(k -> LOG.debug("  - {}", k));
		blackListTest.forEach(k -> map.remove(k));
	}
```

**Comment by jamesward** [comment]
Thanks @bartolom for the detailed feedback.  I know this is all quite a mess.  Sorry for your hassles.

The relative paths across projects is so nasty...  Totally assumes a very specific way of deployment.  So I appreciate you trying to tackle handling that.  Feel free to send any PRs to [webjars-locator-core](https://github.com/webjars/webjars-locator-core).

I'm not too happy about the auto-lower casing we did because it violates one of my WebJar rules by not trusting the upstream metadata.  So I'm happy to revert that despite the downside of duplicated dependencies.  I think a good compromise would be to fix #1452 along with reverting the auto-lower casing change.

**Comment by bartolom** [comment]
I'll try to make a pull request for the additional constructor. But I'll need to check if it solves all our requirements. We now have to cover offline caching with service worker as well. And webjars might surprisingly help with that.
